tests/api/test_addon_uploads.py

```diff
@@ -1922,6 +1953,285 @@ def test_edit_version_both_license_and_custom_license(base_url, session_auth):
+        },
+        data=json.dumps(payload),
+    )
+    print(
+        f'For custom_license "{value}": Response status is '
+        f'{edit_version.status_code}; {edit_version.text}\n'
+    )
+    assert (
+        edit_version.status_code == 400
+    ), f'Actual status code was {edit_version.status_code}'
+    if type(value) is not dict:
+        assert (
+            'You must provide an object of {lang-code:value}.' in edit_version.text
+        ), f'Actual response message was {edit_version.text}'
+    else:
+        assert (
+            'The language code \\"foo\\" is invalid.' in edit_version.text
+        ), f'Actual response message was {edit_version.text}'
+
+
+@pytest.mark.parametrize(
+    'value',
+    [
+        '',
+        None,
+        False,
+        123,
+        'random string',
+    ],
+    ids=[
+        'Empty string',
+        'None/No value',
+        'Boolean',
+        'Integer',
+        'Random string',
+    ],
+)
+@pytest.mark.serial
+@pytest.mark.create_session('api_user')
+def test_edit_version_invalid_compatibility_format(base_url, session_auth, value):
+    """The compatibility field needs to be either a dictionary or a list; other formats should fail"""
+    addon = payloads.edit_addon_details['slug']
+    request = requests.get(
+        url=f'{base_url}{_addon_create}{addon}',
+        headers={'Authorization': f'Session {session_auth}'},
+    )
+    # get the version id of the version we want to edit
+    version = request.json()['current_version']['id']
+    payload = {**payloads.edit_version_details, 'compatibility': value}
+    edit_version = requests.patch(
+        url=f'{base_url}{_addon_create}{addon}/versions/{version}/',
+        headers={
+            'Authorization': f'Session {session_auth}',
+            'Content-Type': 'application/json',
+        },
+        data=json.dumps(payload),
+    )
+    print(
+        f'For compatibility "{value}": Response status is '
+        f'{edit_version.status_code}; {edit_version.text}\n'
+    )
+    assert (
+        edit_version.status_code == 400
+    ), f'Actual status code was {edit_version.status_code}'
+    if value is None:
+        assert (
+            'This field may not be null.' in edit_version.text
+        ), f'Actual response message was {edit_version.text}'
+    else:
+        assert (
+            'Invalid value' in edit_version.text
+        ), f'Actual response message was {edit_version.text}'
+
+
+@pytest.mark.parametrize(
+    'request_value, response_value',
+    [
+        (
+            ['android', 'firefox'],
+            {
+                'android': {'min': '48.0', 'max': '*'},
+                'firefox': {'min': '42.0', 'max': '*'},
+            },
+        ),
+        (['firefox'], {'firefox': {'min': '42.0', 'max': '*'}}),
+        ({'firefox': {'min': '65.0'}}, {'firefox': {'min': '65.0', 'max': '*'}}),
+        ({'android': {'max': '95.0'}}, {'android': {'min': '48.0', 'max': '95.0'}}),
```

Comment:
What is going on here exactly?

Reply:
Yes, the parameters are a bit hard to follow indeed. I've tried to make the `ids` suggestive but I guess that without knowing how the API works it's not all that clear.

So, each parameter contains a valid `compatibility` `request_value` - for example, in the first parameter we send `['android', 'firefox']` as compatibility, which should be interpreted by the API as `compatibility: {'android': {'min': '48.0', 'max': '*'}, 'firefox': {'min': '42.0', 'max': '*'}}`, which is what I've set as the `response_value` parameter to compare against. 

In  the  third parameter, I'm setting the `request_value` as `{'firefox': {'min': '65.0'}}`, which means I'm making my addon compatible with Firefox 65.0 as min value and the API response I'm expecting is `compatibility:  {'firefox': {'min': '65.0', 'max': '*'}}`.

Hope this clarifies things a bit.